File: backend/app/services/sms_service.py
"""
SMS alert service using Twilio.
Falls back gracefully if credentials are not set (logs the message instead).
"""

import logging
from typing import Optional

from app.core.config import settings

log = logging.getLogger(__name__)


async def dispatch_emergency_sms(
    recipients: list[dict],
    vehicle_plate: str,
    coordinates: dict,
    severity: str,
    blood_group: str,
    nearest_hospital: Optional[str],
) -> None:
    """Send emergency SMS to all recipients."""
    if not (settings.twilio_account_sid and settings.twilio_auth_token and settings.twilio_from_number):
        log.info(
            "[SMS] Twilio credentials not configured — SMS simulated:\n"
            f"  TO: {[r.get('phone') for r in recipients]}\n"
            f"  Vehicle: {vehicle_plate} | Severity: {severity} | "
            f"Location: {coordinates} | Blood: {blood_group}"
        )
        return

    try:
        from twilio.rest import Client  # type: ignore

        client = Client(settings.twilio_account_sid, settings.twilio_auth_token)
        lat = coordinates.get("lat", "?")
        lng = coordinates.get("lng", "?")
        body = (
            f"⚠️ ASAAS EMERGENCY ALERT\n"
            f"Vehicle: {vehicle_plate}\n"
            f"Severity: {severity}\n"
            f"Location: https://maps.google.com/?q={lat},{lng}\n"
            f"Blood Group: {blood_group}\n"
            f"Hospital: {nearest_hospital or 'Locating...'}"
        )
        for contact in recipients:
            phone = contact.get("phone")
            if phone:
                message = client.messages.create(
                    body=body,
                    from_=settings.twilio_from_number,
                    to=phone,
                )
                log.info(f"[SMS] Sent to {phone}: SID={message.sid}")
    except Exception as exc:
        log.error(f"[SMS] Error sending SMS: {exc}")

File: server/src/services/smsService.js
// Multi-Channel Emergency SMS Dispatch Gateway
// Supports Fast2SMS (India Gov DLT), Twilio, and Console/Audit Logging

export async function dispatchEmergencySms({ recipients, vehiclePlate, coordinates, severity, bloodGroup, nearestHospital }) {
  const gmapsUrl = `https://maps.google.com/?q=${coordinates.lat},${coordinates.lng}`;
  const message = `[ASAAS CRITICAL ALERT] Vehicle ${vehiclePlate} has suffered a ${severity} collision at ${coordinates.lat.toFixed(4)}N, ${coordinates.lng.toFixed(4)}E. Blood Req: ${bloodGroup || 'Unknown'}. Nearest Facility: ${nearestHospital || 'AIIMS Trauma'}. Live Track: ${gmapsUrl}`;

  const results = [];

  for (const contact of recipients) {
    console.log(`[SMS-GATEWAY] Dispatching Emergency SMS to ${contact.name} (${contact.phone})...`);

    // 1. Check if Fast2SMS API Key is present in environment
    if (process.env.FAST2SMS_API_KEY) {
      try {
        const response = await fetch('https://www.fast2sms.com/dev/bulkV2', {
          method: 'POST',
          headers: {
            'authorization': process.env.FAST2SMS_API_KEY,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            route: 'q',
            message: message,
            numbers: contact.phone.replace(/[^0-9]/g, '')
          })
        });
        const resData = await response.json();
        results.push({ phone: contact.phone, status: 'SENT_FAST2SMS', details: resData });
        continue;
      } catch (err) {
        console.error(`[SMS-GATEWAY] Fast2SMS Error for ${contact.phone}:`, err);
      }
    }

    // 2. Check if Twilio is configured
    if (process.env.TWILIO_ACCOUNT_SID && process.env.TWILIO_AUTH_TOKEN && process.env.TWILIO_PHONE_NUMBER) {
      try {
        const auth = Buffer.from(`${process.env.TWILIO_ACCOUNT_SID}:${process.env.TWILIO_AUTH_TOKEN}`).toString('base64');
        const body = new URLSearchParams({
          From: process.env.TWILIO_PHONE_NUMBER,
          To: contact.phone,
          Body: message
        });

        const response = await fetch(`https://api.twilio.com/2010-04-01/Accounts/${process.env.TWILIO_ACCOUNT_SID}/Messages.json`, {
          method: 'POST',
          headers: {
            'Authorization': `Basic ${auth}`,
            'Content-Type': 'application/x-www-form-urlencoded'
          },
          body: body.toString()
        });
        const twilioData = await response.json();
        results.push({ phone: contact.phone, status: 'SENT_TWILIO', sid: twilioData.sid });
        continue;
      } catch (err) {
        console.error(`[SMS-GATEWAY] Twilio Error for ${contact.phone}:`, err);
      }
    }

    // 3. Fallback: Log to Audit Record & Console
    console.log(`[SMS-GATEWAY] [SIMULATED_CARRIER_DISPATCH] To: ${contact.phone} -> "${message}"`);
    results.push({
      phone: contact.phone,
      status: 'DISPATCHED_SIMULATOR',
      timestamp: new Date().toISOString(),
      preview: message
    });
  }

  return results;
}
